### PR TITLE
AIP-72: Add "XCom" GET endpoint for Execution API

### LINKED ISSUE #43839: AIP-72: Add endpoints for XCom & Variables for Execution subapp
Part of https://github.com/orgs/apache/projects/405/views/1

### PR DESCRIPTION
**Pull request by kaxil**
closes https://github.com/apache/airflow/issues/43839

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/api_fastapi/execution_api/routes/xcoms.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    if not has_xcom_access(key, token):
+        raise HTTPException(
+            status_code=status.HTTP_403_FORBIDDEN,
+            detail={
+                "reason": "access_denied",
+                "message": f"Task does not have access to XCom key '{key}'",
+            },
+        )
+
+    query = BaseXCom.get_many(
```
Comment (ashb) [inline airflow/api_fastapi/execution_api/routes/xcoms.py:71]: We shouldn't use get_many

Comment (kaxil) [reply airflow/api_fastapi/execution_api/routes/xcoms.py:71]: Added comment on why:

https://github.com/apache/airflow/blob/12f1cf2e8232146f534b915ad2ff5f0a456347da/airflow/api_fastapi/execution_api/routes/xcoms.py#L66-L70

Comment (ashb) [reply airflow/api_fastapi/execution_api/routes/xcoms.py:71]: Oh yes, _that_.

File: airflow/api_fastapi/execution_api/routes/__init__.py [hunk 2]
```diff
@@ -17,10 +17,11 @@
...
 from airflow.api_fastapi.common.router import AirflowRouter
-from airflow.api_fastapi.execution_api.routes import connections, health, task_instance, variables
+from airflow.api_fastapi.execution_api.routes import connections, health, task_instance, variables, xcoms
 
 execution_api_router = AirflowRouter()
 execution_api_router.include_router(connections.router, prefix="/connections", tags=["Connections"])
 execution_api_router.include_router(health.router, tags=["Health"])
 execution_api_router.include_router(task_instance.router, prefix="/task_instance", tags=["Task Instance"])
 execution_api_router.include_router(variables.router, prefix="/variables", tags=["Variables"])
+execution_api_router.include_router(xcoms.router, prefix="/xcoms", tags=["XComs"])
```
Comment (ashb) [inline airflow/api_fastapi/execution_api/routes/__init__.py]: We should be consistent on our pluralization - task_instance is singular, all the rest are plural

File: airflow/api_fastapi/execution_api/routes/xcoms.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+                "reason": "not_found",
+                "message": f"XCom with key '{key}' not found for task '{task_id}' in DAG '{dag_id}'",
+            },
+        )
+
+    try:
+        xcom_value = BaseXCom.deserialize_value(result)
+    except json.JSONDecodeError:
+        raise HTTPException(
+            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
```
Comment (ashb) [inline airflow/api_fastapi/execution_api/routes/xcoms.py]: 4xx is wrong error code here, this error isn't result of a client error. I guess 500 is the only suitable thing here.

**Review by ashb — APPROVED** [review]
Few nits, LGTM otherwise

File: airflow/api_fastapi/execution_api/routes/xcoms.py (hunk 3 above)
Comment (kaxil) [reply airflow/api_fastapi/execution_api/routes/xcoms.py]: 👍  will change it. My initial thought was they (user of the client), might delete/change that entry somehow coz it isn't going to work. But I agree 500 is more apt

File: airflow/api_fastapi/execution_api/routes/__init__.py (hunk 2 above)
Comment (kaxil) [reply airflow/api_fastapi/execution_api/routes/__init__.py]: Yeah I went back and forth on this. Reasoning was that `task_instance` was always ever going to operator on a single TI, while the other endpoints would operate on one or more of it (XCom, vars, conns etc).

I will make them singular for now.

Comment (ashb) [reply airflow/api_fastapi/execution_api/routes/__init__.py]: The single vs plural question in API design is one as old as REST.

Comment (kaxil) [reply airflow/api_fastapi/execution_api/routes/__init__.py]: :) -- I changed my mind again -- going to make it plural — so I don’t have to change it in future if we add list, collection get endpoints and the fact that Public API has it plural too, so for code consistency!

Comment (kaxil) [reply airflow/api_fastapi/execution_api/routes/__init__.py]: done in https://github.com/apache/airflow/pull/43894/commits/542a7034e18bbf35d1b7b6cc3cb538ad2ad0322b

**Pull request merged**